From 90779d092c9ba5a7b39fc38cc64d5702172c8a48 Mon Sep 17 00:00:00 2001
From: Robert Dobrowolski <robert.dobrowolski@intel.com>
Date: Wed, 20 Jan 2016 12:34:12 +0100
Subject: [PATCH] usb: gadget: Fix mtp_read freeze when switching device config
 to rndis and then back to mtp

When switching config to rndis,acm,adb, mtp_read should return when
device is offline, instead its stuck on wait_event_interruptible,
waiting for device to be online. Its because mtp device is always
offline when config does not include mtp mode. Because of that process
is put to sleep and usb_mtp driver is not released as expected, this
prevents MtpService from reopening usb_mtp when configuration is
switched from rndis to mtp.

Change-Id: Ia430e4b63f587c3c40fb86808033a7ad45c6dbc0
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-12506
Signed-off-by: Robert Dobrowolski <robert.dobrowolski@intel.com>
Reviewed-on: https://android.intel.com:443/460312
---
 drivers/usb/gadget/f_mtp.c | 11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

diff --git a/drivers/usb/gadget/f_mtp.c b/drivers/usb/gadget/f_mtp.c
index d58d010..31b5cf1 100644
--- a/drivers/usb/gadget/f_mtp.c
+++ b/drivers/usb/gadget/f_mtp.c
@@ -564,14 +564,11 @@ static ssize_t mtp_read(struct file *fp, char __user *buf,
 
 	DBG(cdev, "mtp_read(%zu)\n", count);
 
-	/* we will block until we're online */
-	DBG(cdev, "mtp_read: waiting for online state\n");
-	ret = wait_event_interruptible(dev->read_wq,
-		dev->state != STATE_OFFLINE);
-	if (ret < 0) {
-		r = ret;
-		goto done;
+	if (dev->state == STATE_OFFLINE) {
+		DBG(cdev, "mtp_read: state offline, return\n");
+		return -EIO;
 	}
+
 	spin_lock_irq(&dev->lock);
 	if (dev->ep_out->desc) {
 		len = usb_ep_align_maybe(cdev->gadget, dev->ep_out, count);
-- 
1.9.1

